### PR TITLE
Support Exporting Contest Results

### LINKED ISSUE #307: Support Exporting Contest Results
csv

### PR DESCRIPTION
**Pull request by twd2**
Fix #307

### CODE REVIEW

File: vj4/ui/templates/contest_status.html [hunk 1]
```diff
@@ -3,6 +3,7 @@
 <div class="row"><div class="medium-12 columns">
   <div class="section visible">
     <div class="section__body no-padding">
+      <a href="{{ reverse_url('contest_status_download', tid=tdoc['doc_id']) }}">{{ _('Download') }}</a>
```
Comment (twd2) [inline vj4/ui/templates/contest_status.html]: @breeswish

**Comment by breeswish** [comment]
I guess opening CSV in Excel will be crap if there are Chinese characters in the user name...

**Comment by twd2** [comment]
@breeswish  http://blog.csdn.net/leonzhouwei/article/details/8447643

**Comment by breeswish** [comment]
@twd2 We need to export in GBK charset.

File: vj4/ui/templates/partials/contest_status_download_acm.csv [hunk 2]
```diff
@@ -0,0 +1,6 @@
+{{ _('Rank') }},{{ _('User') }},{{ _('Solve') }},{{ _('Time') }},{% for pid in tdoc['pids'] %}#{{ loop.index }} {{ pdict[pid]['title'] }},{% endfor %}
```
Comment (breeswish) [inline vj4/ui/templates/partials/contest_status_download_acm.csv]: I think there is no need to use template here. Current template is extremely hard to read.

File: vj4/handler/contest.py [hunk 3]
```diff
@@ -287,6 +287,25 @@ class ContestStatusHandler(base.Handler, ContestStatusMixin):
...
+  @base.sanitize
+  async def get(self, *, tid: objectid.ObjectId):
+    tdoc, tsdocs = await contest.get_and_list_status(self.domain_id, tid)
+    if (not contest.RULES[tdoc['rule']].show_func(tdoc, self.now)
+        and not self.has_perm(builtin.PERM_VIEW_CONTEST_HIDDEN_STATUS)):
+      raise error.ContestStatusHiddenError()
+    udict, pdict = await asyncio.gather(user.get_dict([tsdoc['uid'] for tsdoc in tsdocs]),
+                                        problem.get_dict(self.domain_id, tdoc['pids']))
+    ranked_tsdocs = contest.RULES[tdoc['rule']].rank_func(tsdocs)
+    self.render('contest_status_download.csv', content_type='test/csv',
```
Comment (breeswish) [inline vj4/handler/contest.py]: The downloaded file name is `download`, which is not friendly.

**Comment by twd2** [comment]
We should use *Unicode*.

**Comment by breeswish** [comment]
For text editor users, most of the editors can recognize GBK well so it is not a problem.
For excel users, it only accepts GBK as a valid charset.

**Comment by twd2** [comment]
@breeswish  merge me

**Pull request merged**